### PR TITLE
Add getCollateralToken rpc

### PR DESCRIPTION
**Pull request by jingyi2811**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Add getCollateralToken rpc

#### Which issue(s) does this PR fixes?:

Fixes part of #488

### CODE REVIEW

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (Jouzo) [inline packages/jellyfish-api-core/src/category/loan.ts:59]: Maybe `CollateralTokenResult` to know that it's an interface returned from a call

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (Jouzo) [inline packages/jellyfish-api-core/src/category/loan.ts:36]:
```suggestion
   * @return {Promise<CollateralToken>} CollateralToken
```

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (Jouzo) [inline packages/jellyfish-api-core/src/category/loan.ts:39]:
```suggestion
    return await this.client.call('getcollateraltoken', [getCollateralToken], { factor: 'bignumber' })
```

**Review by Jouzo — CHANGES_REQUESTED** [review]
Need to handle `CollateralToken` `factor` property as BigNumber

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:121]: is this needed??

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (Jouzo) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:121]:
```suggestion
  it('should getCollateralToken as empty object if height < activateAfterBlock', async () => {
```
Or document somewhere that the test is expected to return an empty result

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:121]: I am going to change this to `should getCollateralToken as empty object`
i do not add `if height < activateAfterBlock` because the describe name has similar text

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:76]: Good catch. I am going to change this to
`return await this.client.call('getcollateraltoken', [collateralToken], 'bignumber')`

Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts]: I choose this:` the interface is documented as BigNumber but the call will default to 'number' for activateAfterBlock`

Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts]: Hmm, 

1. factor: 'bignumber', 
2. factor: 'bignumber', activateAfterBlock: 'bignumber'

Not sure why, changing the code to these 2 does not provide result that works as expected. Both still returns output that has number (Not BigNumber)

Therefore, I stick to `await this.client.call('getcollateraltoken', [collateralToken], 'bignumber')`

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (canonbrother) [inline packages/jellyfish-api-core/src/category/loan.ts:100]:
```suggestion
  token?: string
  height?: number
```

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:59]: can you move the `describe` into `it`??

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:59]: I need another describe to reset the block and don't want to have side effect that may cause by 
https://github.com/DeFiCh/jellyfish/pull/546/files#diff-d66e7acf80be154ffa1d4c2fc74b1cd3bb3cbeb06eea50e9e542df1585a1a422R31

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:100]: You are right

File: website/docs/jellyfish/api/loan.md
Comment (jingyi2811) [inline website/docs/jellyfish/api/loan.md:162]: Both variables have the same elements. Hence, I refactor the code.

**Comment by jingyi2811** [comment]
Kindly ignore any checking on any string from now, where it should not be an empty string or more than 8 chars. There are many RPCs still not handle this well and throw other unrelated error

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (Jouzo) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: if you call getCollateralToken without parameters, it return a list of all activated token ?
Seems weird, it should always return a single token and list should accept a height parameter or activated boolean.

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: getcollateraltoken will return all current collateral tokens. If there is none then it will be empty list. listcollateraltokens will return all txs over time

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: Well hahahahaha

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: @fuxingloh

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: if you call getCollateralToken without parameters, it return a list of all activated token ? => Actually, the best is C++ programmer create another rpc for this, maybe called listCurrentCollateralTokens and for listCollateralTokens, change it to listAllCollateralTokens

Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]:
```suggestion
      const data = await testing.rpc.loan.getCollateralToken()
```

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: I adopted your suggestion. And changed `collateralToken: GetCollateralToken` to `collateralToken?: GetCollateralToken` to fix the compilation error raised from that 

Screenshot 2021-09-10 at 11 16 33 AM

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:239]: Test run failed. To fix the test, I change the code to  
Screenshot 2021-09-10 at 11 27 31 AM

File: website/docs/jellyfish/api/loan.md
Comment (surangap) [inline website/docs/jellyfish/api/loan.md:155]: docs?
```suggestion
```

File: website/docs/jellyfish/api/loan.md
Comment (surangap) [reply website/docs/jellyfish/api/loan.md:164]: Can change to following. gives more suitable name for both RPCs
```suggestion
interface CollateralTokenDetails {
```

File: website/docs/jellyfish/api/loan.md
Comment (jingyi2811) [reply website/docs/jellyfish/api/loan.md]: Hahaha. Fixed

File: website/docs/jellyfish/api/loan.md
Comment (jingyi2811) [reply website/docs/jellyfish/api/loan.md]: Fixed

**Comment by jingyi2811** [comment]
Don't approve this PR first. I need to resolve merge conflicts after createVault rpc is merged to main

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:58]:
```suggestion
  it('should getCollateralToken with empty string if the query height is below the activation block height', async () => {
```

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:63]:
```suggestion
  it('should getCollateralToken if the height is after the activation block height', async () => {
```

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:171]:
```suggestion
  it('should getCollateralToken with empty string if the height is below the activation block height', async () => {
```

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:176]:
```suggestion
  it('should getCollateralToken if the height is after the activation block height', async () => {
```

**Review by surangap — COMMENTED** [review]
looks good. Requesting for minor test name refactors.

**Review by surangap — APPROVED** [review]

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:76]: kindly refactor the test
```
describe(getCollateralToken)
  it(should getCollateralToken)
  it(should getCollateralToken with token only)
  it(should ... with height only)
  // error test
```

**Review by canonbrother — CHANGES_REQUESTED** [review]

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:76]: They need to be in different describe to reset the block

Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:76]: nope

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:76]: I think you assume activateAfterBlock to be larger if the next `it`

Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:76]: more behaviour to document
create 3 collateral token
1. TSLA, 120
2. AAPL, 140
3. AMZN, 160
get(TSLA) -> return TSLA
get(140) -> return [AAPL, AMZN]
get(AMZN, 120) -> return AMZN

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (ivan-zynesis) [inline packages/jellyfish-api-core/src/category/loan.ts:124]: Does "id" work?

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (ivan-zynesis) [inline packages/jellyfish-api-core/src/category/loan.ts:125]:
```suggestion
   * @param {number} [collateralToken.height] height filter, activation height less than or equal to
```

**Review by ivan-zynesis — CHANGES_REQUESTED** [review]

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (surangap) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts:76]: @jingyi2811 i think what @canonbrother want to say is, you can setup something like
```
TSLA, 120
AAPL, 140
AMZN, 160
```
inside `beforeAll()` and then write all the test cases for `it(should getCollateralToken), it(should getCollateralToken with token only), it(should ... with height only)` within the same container. I think it's doable. Just make sure tests are independent. 
But you still have to keep `Loan getCollateralToken with no parameter` in a separate container since it needs block height adjustment and it's clearer that way alone.

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts]: The comment text is copied from Ain. Id does not work here. I have changed the text to `Symbol of collateral token`

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:125]: The text is copied from ain. I stick to `Valid at specified height`. If want to change, the text should be height filter, `activation height more than or equal to`

**Comment by jingyi2811** [comment]
I convert this PR to draft. Once I open this back, will text you all again

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts]: Okay refactor the code again to use 2 describes. I draw a conclusion that the number the describes you have depends on how many times you wrote `await testing.container.waitForBlockHeight(120)` [Not too sure how true is this formula]

one APPL is good enough, no need the other 2

This is how my new code looks like. 

```
describe('Loan getCollateralToken', () => {
describe('Loan getCollateralToken with parameters token or height', () => {
   describe('Loan getCollateralToken with parameter token and height', () => {
   describe('Loan getCollateralToken with parameter height only', () => {
   describe('Loan getCollateralToken with parameter token only', () => {
}   
```

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts]: I suppose to be able to think that another 2 describes. Ya keep everything in 1 describe is impossible but not 4 describes.

File: packages/jellyfish-api-core/src/category/loan.ts
Comment (surangap) [reply packages/jellyfish-api-core/src/category/loan.ts:125]: looking at c++ side implementation looks like `height` has a default value, which is the current block height. so if a  height is not given, the default value will be used. I'm thinking if we should reflect that in the documentation🤔 or just keep it as it is. 
`current block height` is not a constant value, Then how should we do it?
`@param {number} [collateralToken.height = current block height]`
@fuxingloh

Comment (surangap) [reply packages/jellyfish-api-core/src/category/loan.ts:125]: Ref - https://github.com/DeFiCh/ain/blob/3b17b2ede136fa1931e2831415ddf7e5f4a15a22/src/masternodes/rpc_loan.cpp#L187

Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:125]: Or set the default value for height in Javascript side?
https://github.com/DeFiCh/jellyfish/pull/568#discussion_r705105016

Comment (fuxingloh) [reply packages/jellyfish-api-core/src/category/loan.ts:125]: > Or set the default value for height in Javascript side?
> [#568 (comment)](https://github.com/DeFiCh/jellyfish/pull/568#discussion_r705105016)

we can't really set a default value for this since it's stateful dependent on current block height,

```
@param {number} [collateralToken.height = CurrentBlockheight]
```

as @surangap comment, the above works

Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:125]: Fixed

**Review by surangap — DISMISSED** [review]

File: packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts
Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts]: there still got duplicate tests
```
// eg: token not exists
get({token: nonexistence })
// is same as
get({ token: nonexistence: height: 200})
```

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts]: Is there any other duplicate tests too? I want to kill 2 birds with 1 stone (Meaning Fix 2 things in 1 attempt) :)

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/getCollateralToken.test.ts]: Fixed

**Review by surangap — APPROVED** [review]

**Review by ivan-zynesis — APPROVED** [review]

**Comment by canonbrother** [comment]
FYI, lesser `describe`, cleaner..  not critical.. i ignore it first
the test requires to 2 containers to test 1 for height manipulation and 1 for filter manipulation ( the getCollateral is actually a listCollateral here)
```
// filter manipulation
describe('Loan getCollateralToken with filters')
  it('should getCollateralToken with token and height)
  it('should getCollateralToken with token)
  it('should getCollateralToken with height)
  it('should not .... as token is not exists')
  it('should not .... as height below activateAfterBlock')
```

**Review by canonbrother — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**
